Audio Decomp'd - ROM is OK (#1236)
zeldaret/oot: docs/Music Macro Language.md

A comment always starts with a # character.  This may appear anywhere in a line, and all text after the # will be ignored by editors or parsers.  These comments will not be saved or stored in the binary representation of a sequence.  They are purely for documentation purposes only.

Metainstructions are instructions that are read by the editor or parser, but do not get stored in the resultant binary file directly.  The following metainstructions are supported by the parser provided by `assemble_sequence`.  Other tools may support additional metainstructions, but these are always supported by all tools.

**.define** *symbol* *value*

**Description:** Defines a symbol that may be used elsewhere in the sequence script.  The symbol may consist of any Unicode letter, number, or underscores.  The value must be a primitive value.  Symbols must be defined only once in a sequence, similar to 

**Parameters:**

* `symbol` - The symbol being defined in the sequence.

> * `value` - A primitive that will the symbol represents.

Reviewer: Applied in #1509.

.define IO_PORT_MAIN 0

# Binary values
.define TRUE  1
.define FALSE 0
```

**alternate.mus.inc**
```
.desc title Alternate Melody

.channel channel0_alt
... various instructions here...
end
```

**Label Metainstructions**

The following metainstructions allow you to define a label that points to the relevant set of instructions.  These metainstructions are the following:

| Metacommand | Description                                                 |
| ----------- | ----------------------------------------------------------- |
| .sequence   | A series of sequence instructions within a track section.   |
| .channel    | A series of channel instructions within a track section.    |
| .layer      | A series of note layer instructions within a track section. |
| .table      | A table of entries or bytes within a data section.          |
| .array      | An array of bytes or shorts in a data section.              |
| .filter     | References a filter structure in a data section.            |
| .envelope   | References an envelope structure in a data section.         |
| .buffer     | References a buffer within a data section.                  |

**Section Metainstructions**

> There are metainstructions specific to certain sections that will be documented in those sections specifically.  These will be described in those sections as appropriate.  Note that metainstructions are always distinguished by starting with a period, whereas instructions do not.

Reviewer: Applied in #1509.

#### Sequence Subsections

##### Sequence Registers

Certain instructions support loading or storing values and utilize a register associated to the sequence for this task.  For documentation purposes, this is referred to as the `$V` register.  This register is one byte in size; most instructions treat it as a signed value.

##### IO Ports

Each sequence player contains 8 IO ports that are one byte in size each.  These ports can be used to communicate state information between other parts of the game engine, or across different parts of the sequence.  For example, a sequence script could use an IO port that, when set to a certain value, changes what instrument is used.

##### Metainstructions

**.sequence** *label name*

> **Description:** Defines a label for the following subsection of sequence instructions.  Note that it is considered an error if a sequence label is proceeded with anything that is not a sequence command.

Reviewer: Applied in #1509.

**testchan** *N*

**Description:** Stores whether or not channel *N* is currently active in `$V`.  If the channel is active, `$V` will equal `0`, otherwise it will equal `1`.

**Parameters:**
* `N` [integer: 0-15] - The channel number to test.

**Example:**
```
.sequence wait_until_done
delay1
testchan 1
beqz wait_until_done
# Instructions here only execute once channel 1 reaches an end command.
```

**stopchan** *N*

**Description:** Stops channel *N* from executing any further instructions.  Any notes currently playing on the channel will cease.

**Parameters:**
* `N` [integer: 0-15] - The channel number to stop.

**Example:**
```
ldchan 0, chan0
ldchan 1, chan1
delay 50
stopchan 1      # Stops channel 1 after 50 ticks, even if there's additional instructions.
```

**subio** *N*

**Description:** Subtracts from `$V` whatever value is in IO port *N*.

**Parameters:**
* `N` [integer: 0-7] - The IO port to read.

**Example:**
```
# Assume that the current value in $V is 5 and the current value in IO port 1 is -5.
subio 1
stio 0
# IO port 0 now contains the current value of $V, which is 0.
```

**ldres** *IO*, *type*, *index*

> **Description:** Loads an audio resource into memory asynchronously.  The type of resource is specified via *type*, where 0 is a sequence, 1 is a soundfont, and 2 is a sample bank.  The IO parameter is used to track the status of the loading process.  The audio subsystem will update this value to 0 once the resource has been loaded.  Note that the IO port doesn't change this value to anything but 0 first, so it's recommended to set the IO port to a non-zero value prior to using `ldres`.

Reviewer 2: "audio resource" is the first time I've heard that term used to describe the collection of data that is `sequence`/`soundfont`/`sample bank`. It does make sense to me, just highlighting that nothing in decomp uses that term (which could change)

Reviewer: TBD CC @MNGoldenEagle @Dragorn421

**Parameters:**
* `val` [integer: 0-255] - Value to subtract

**Example:**
```
ldio 3   # Load value from IO port 3 into $V
sub 2    # $V -= 2
stio 3   # Store $V to IO port 3
```

**and** *val*

**Description:** Performs a bitwise and operation between the value stored in the `$V` register and the value specified.

**Parameters:**
* `val` [integer: 0-255] - Value to perform a bitwise and with

**Example:**
```
ldio 3     # Load value from IO port 3 into $V
and 0xF0   # Set lower four bits to 0
stio 2     # Store resulting value into IO port 2
```

**ldi** *val*

**Description:** Sets the `$V` register to the provided integer.

**Parameters:**
* `val` [integer: 0-255] - Value to load into the register

**Example:**
```
ldi 2    # $V = 2
stio 0   # Set IO port 0 to 2
```

**dyncall** *table*

**Description:** Makes a function call using an entry from the specified table in a data section.  The value in the `$V` register is treated as the index to look up within the table.  Once an `end` command is reached in the called function, the player will return back and execute the next instruction after this one.

> If `$V` is currently set to -1 or the call depth is at 3, then the function call will fail and be skipped.  Note that the sequence player does not perform any bounds checking on `$V`; if it's larger than the table, the sequence player will jump to a potentially unexpected location in the script and could result in undefined behavior.

Reviewer 2: What does the `dyn` mean here? I suppose it's tied to all the `dynTable` commands to indicate they're all in a group? (noting decomp does also use the `dyn` terminology)

Reviewer: TBD.

notevg PITCH_G4, 6, 0
notevg PITCH_C5, 6, 0

.data

.array shortgates
.byte 20
.byte 50
.byte 80
.byte 100
.byte 120
.byte 150
.byte 180
```

**ldshortvelarr** *array*

**Description:** Loads the specified byte array of note velocity values into the sequence player.  This array can be used in the note layer by the `ldshortvel` command.  This allows you to change velocity values for consecutive notes with only a single byte, however only up to 16 values can be read from the array.

Note that the short note velocity array applies across all channels and note layers in the sequence player.  If the array is changed at any point, this will also change the velocity array for all note layers simultaneously.  This will not affect the velocity values already loaded.

If this command is not used but the `ldshortvel` is used, the velocity value will be read from `gDefaultShortNoteVelocityTable`, which consists of the following values:

`{ 12, 25, 38, 51, 57, 64, 71, 76, 83, 89, 96, 102, 109, 115, 121, 127 }`

**Parameters:**
* `array` [array label] - Label to an array in a data section containing up to sixteen bytes.

**Example:**
```

> ldshortvelarr shortvels    # Assigns the array shortgates to the player's short note lookup table

Reviewer: Applied in #1509.

ldchan 0, chan0            # Loads channel 0
delay 500
end

.channel chan0
short                      # Activates short note mode on this channel
ldlayer layer0             # Loads note layer 0 for this channel
delay 500
end

.layer layer0

> ldshortvel 5               # Loads gate value at index 5 (27) to be used for subsequent notes.

Reviewer: Applied in #1509.

notedvg PITCH_C4, 6, 0     # Plays C4 with gate 27 and duration of 6 ticks.
notevg PITCH_D4, 6, 0      # Plays D4 with gate 27 and the same duration as previous.
notevg PITCH_F4, 6, 0

> ldshortgate 3              # Loads gate value at index 3 (10) to be used for subsequent notes.

Reviewer 3: I assume this was meant to be (`ldshortvelarr` example)
```suggestion
ldshortvel 3               # Loads velocity value at index 3 (10) to be used for subsequent notes.
```

notevg PITCH_G4, 6, 0      # Plays G4 with gate 10 and same duration as previous.
notevg PITCH_G4, 6, 0
notevg PITCH_C5, 6, 0

.data

> .array shortvel

Reviewer 3: ```suggestion
.array shortvels
```

Reviewer: Applied in #1509.

Once mute has been activated, it cannot be disabled again via command.

**Parameters:** None.

**Example:**
```
ldchan 1, chan1  # Starts playing channel 1.
delay 500
mute             # Mutes the sequence.
delay 40
end
```

**mutescale** *numerator*

**Description:** Sets the fractional value that should be applied to the sequence's current volume when mute mode is enabled.  The value passed in the parameter will be divided by 127 and the result stored in the sequence player's configuration.  Unlike `mutebhv`, this can be used at any point in the sequence to change the mute scale factor.

For example, if set to 127 and the sequence is muted, the volume will not reduce at all.  If set to 80, the volume will reduce by 63%.  If set to 0, the volume will mute entirely.  Using a value larger than 127 will result in undefined behavior, as this will result in a negative scale factor.  By default, this will be set to 100%.

This scaling factor is only used if flag 0x20 is set in the sequence player's mute behavior.

**Parameters:**
* `numerator` [integer: 0-127] - The number to divide by 127

**Example:**
```
mutebhv 0x20        # Sets mute behavior to reduce sequence volume when muted

> mutescale 40        # Reduce volume by 31.5% when muted

Reviewer 3: ```suggestion
mutescale 40        # Scale volume by 31.5% when muted
```
"reduce by 30%" would be "scale by 70%"

ldchan 1, chan1     # Play channel 1 at 100% volume
delay 500
mute                # Channel 1 continues playing at 31.5% volume
```

**initchan** *mask*

> **Description:** Initializes the indicated channels on the sequence player and prepares them for playing audio.  This configure's the channels with the default soundfont, assigned mute behavior, and configured note allocation policy.  Each bit on the bitmask represents a different channel, with the most significant bit representing channel 16 and the least significant bit representing channel 1.

Reviewer 3: ```suggestion
**Description:** Initializes the indicated channels on the sequence player and prepares them for playing audio.  This configures the channels with the default soundfont, assigned mute behavior, and configured note allocation policy.  Each bit on the bitmask represents a different channel, with the most significant bit representing channel 16 and the least significant bit representing channel 1.
```

Reviewer: Applied in #1509.

> It is not strictly necessary to use `initchan` for the channels to be used, however the previous mentioned settings will not be configured on the channel and will be in an undefined state.  For obvious reasons, this is not recommended.

Reviewer 3: ```suggestion
It is not strictly necessary to use `initchan` for the channels to be used, however the previously mentioned settings will not be configured on the channel and will be in an undefined state.  For obvious reasons, this is not recommended.
```

Reviewer: Applied in #1509.

Sequence transpositions are ignored by drum and sound effects.

**Parameters:**
* `transposition` [integer: -128-127] - The number of semitones to transpose by

**Example:**
```
transpose 10   # Transposes the notes played up by 10 semitones
delay 50
```

**freenotelist**

**Description:** Frees the notes reserved on the sequence player and returns them back to the global note pool.

It's recommended to avoid using this command while audio is playing for the sequence, as it can cause undefined behavior.

**Parameters:** None.

**Example:**
```
ldchan 1, chan1
ldchan 9, chan9
delay 1500
freenotelist      # Frees any reserved notes on the sequence player.
```

**allocnotelist** *count*

**Description:** Frees any reserved notes assigned to the sequence player, then reserves the number of notes requested for the sequence player.  The notes in this pool can then be used by channels and note layers when needed, depending on the note allocation policy set.

It's recommended to avoid using this command while audio is playing on the sequence, as this can result in undefined behavior.

> In general, it is not necessary to use this command, as by default the sequence player will use any free notes in the global note pool.  This command also is not useful for addressing issues with notes cutting out on channels in a sequence, as this applies to the sequence player itself.  Howver, it can be useful for preventing other sequences from stealing notes from this sequence.

Reviewer 3: ```suggestion
In general, it is not necessary to use this command, as by default the sequence player will use any free notes in the global note pool.  This command also is not useful for addressing issues with notes cutting out on channels in a sequence, as this applies to the sequence player itself.  However, it can be useful for preventing other sequences from stealing notes from this sequence.
```

Reviewer: Applied in #1509.

| 0B | cdelay | sampleptr | ldchan | stcio | ldcio | subio | ldio | rldlayer | ldlayer   | dynldlayer |    | ptradd      | ldseq        | transpose   | fontinstr    | *jump         |
| 0C | cdelay | sampleptr | ldchan | stcio | ldcio | subio | ldio | rldlayer | ldlayer   | dynldlayer |    | randptr     | ldi          | panweight   | reset        | *call         |
| 0D | cdelay | sampleptr | ldchan | stcio | ldcio | subio | ldio | rldlayer | ldlayer   | dynldlayer |    | instr       | stopchan     | pan         | gain         | *delay        |
| 0E | cdelay | sampleptr | ldchan | stcio | ldcio | subio | ldio | rldlayer | ldlayer   | dynldlayer |    |             | lditoptr     | preqscale   | bendfine     | *delay1       |
| 0F | cdelay | sampleptr | ldchan | stcio | ldcio | subio | ldio | rldlayer | ldlayer   | dynldlayer |    |             | stptrtoseq   | vol         |              | *end          |

Instructions marked with an asterisk (\*) are documented in the section *Branching Instructions*.

**cdelay** *ticks*

> **Description:** Delays processing of the channel script by the provided number of ticks (up to 16).  A delay of 0 is effectively a no-op.

Reviewer 2: why `c` in `cdelay`? Does that mean `channel`? Would `chandelay` make more sense for consistency? Is that `c` the same in `stcio` and `ldcio`?

Reviewer: TBD.
